### PR TITLE
Add new option useSecureParse, move secure parsing under it and add normal parsing as default

### LINKED ISSUE #650: Remove secure-json-parse from Gateway
I think we could safely consider the target services as trusted and remove the secure-json-parse call in https://github.com/mercurius-js/mercurius/blob/25500ddf95f3cfafaab23d03248ac6c52740142c/lib/gateway/request.js#L92.

Even better, we could put it behind an option.

### PR DESCRIPTION
**Pull request by radomird**
Closes #650

### CODE REVIEW

**Comment by mcollina** [comment]
Can you please add a unit test? You can also use proxyquire.

**Comment by radomird** [comment]
> Can you please add a unit test? You can also use proxyquire.

Yes of course - still in progress :)

**Comment by radomird** [comment]
Ready for review.

File: lib/gateway/request.js [hunk 1]
```diff
@@ -89,7 +89,12 @@ function sendRequest (request, url) {
           }
 
           try {
-            const json = sJSON.parse(data.toString())
+            let json
+            if (useSecureParse) {
+              json = sJSON.parse(data.toString())
+            } else {
+              json = JSON.parse(data.toString())
+            }
```
Comment (simoneb) [inline lib/gateway/request.js:97]: nit

```suggestion
            const json ? (useSecureParse ? sJSON : JSON).parse(data.toString())
```

File: test/gateway/request.js [hunk 2]
```diff
@@ -84,3 +84,82 @@ test('sendRequest method rejects when response contains errors', async (t) => {
...
+    query ServiceInfo {
+      _service {
+        sdl
+      }
+    }
+    `
+    })
+  })
+
+  // cecking for prototype leakage: https://github.com/fastify/secure-json-parse#introduction
```
Comment (simoneb) [inline test/gateway/request.js:120]:
```suggestion
  // checking for prototype leakage: https://github.com/fastify/secure-json-parse#introduction
```

File: test/gateway/request.js [hunk 3]
```diff
@@ -84,3 +84,82 @@ test('sendRequest method rejects when response contains errors', async (t) => {
...
+  )({
+    method: 'POST',
+    body: JSON.stringify({
+      query: `
+    query ServiceInfo {
+      _service {
+        sdl
+      }
+    }
+    `
```
Comment (simoneb) [inline test/gateway/request.js:116]: I'm not sure I understand why we're sending a request body. what are we doing with it?

Comment (radomird) [reply test/gateway/request.js:116]: Actually nothing, I copied the previous test and forgot to remove it. Will do it now.

File: lib/gateway/request.js (hunk 1 above)
Comment (radomird) [reply lib/gateway/request.js:97]: Did you mean this maybe:
```js
const json = (useSecureParse ? sJSON : JSON).parse(data.toString())
```

File: test/gateway/request.js (hunk 3 above)
Comment (simoneb) [reply test/gateway/request.js:116]: ok but they I'm wondering, what are we testing here? the behavior you're testing is on a plain route, not a graphql route. I'm confused

File: lib/gateway/request.js (hunk 1 above)
Comment (simoneb) [reply lib/gateway/request.js:97]: yes

File: test/gateway/request.js (hunk 3 above)
Comment (radomird) [reply test/gateway/request.js:116]: We're testing if the response is parsed correctly with useSecureParse flag set to true.
If you look at the response you'll see it's a stringified object containing __proto__ the goal is to make sure that the secure parsing still works.

Comment (simoneb) [reply test/gateway/request.js:116]: what I don't understand is why this behavior is affecting plain HTTP requests. this is mercurius, why are we implementing behaviors which affect requests that should not be affected outside of what mercurius does?

Comment (radomird) [reply test/gateway/request.js:116]: The documentation is not very clear about this, but this article sheds a bit more light on the topic: https://www.nearform.com/blog/implementing-a-graphql-gateway-with-fastify/

The idea is that `mercurius` can work in a so called *"gateway mode"*. In that mode the server is connecting to different services, after it gets the  results of those services, it combines the output of all of them into one final schema. Now, part of the gateway is the `request.js` , that Matteo asked to be changed, and the reasoning behind is I think clear from the ticket.

Comment (simoneb) [reply test/gateway/request.js:116]: thanks, makes sense now!

**Review by simoneb — APPROVED** [review]
LGTM

**Comment by simoneb** [comment]
@radomird doesn't the new option need to be documented?

**Comment by radomird** [comment]
> @radomird doesn't the new option need to be documented?

I does, will updated the docs.

**Review by simoneb — APPROVED** [review]
Perfect 🚀

**Review by mcollina — APPROVED** [review]
lgtm

**Pull request merged**